File: analytics/anomaly2/phish/anomaly2.py
#!/usr/local/bin/python

# MINNOW anomaly2
# anomaly detection from unbounded key range
# print found anomalies to screen

import sys
from my_double_linked_list import MyDoubleLinkedList
import phish

# command-line defaults

nthresh = 24
mthresh = 4
maxactive = 3*131072

# packet stats

nrecv = 0
nunique = 0

# anomaly stats

ptrue = 0
pfalse = 0
nfalse = 0
ntrue = 0

# key stats

nactive = 0
nunique = 0

# one active key

class Akey():
  def __init__(self):
    self.key = 0
    self.count = 0
    self.value = 0
    self.prev = None
    self.next = None

# key-value hash table for received datums

kv = {}

# parse command line
# optional arguments:
#   -t = check key for anomalous when appears this many times
#   -m = key is anomalous if value=1 <= this many times
#   -a = size of active key set

def options(args):
  global nthresh,mthresh,maxactive
  iarg = 1
  while iarg < len(args):
    if args[iarg] == "-t":
      nthresh = int(args[iarg+1])
      iarg += 2
    elif args[iarg] == "-m":
      mthresh = int(args[iarg+1])
      iarg += 2
    elif args[iarg] == "-a":
      maxactive = int(args[iarg+1])
      iarg += 2
    else: break

  if iarg != len(args):
    print "Syntax: anomaly2.py options"
    sys.exit()

  if nthresh <= 0 or mthresh <= 0 or maxactive <= 0:
    print "ERROR: invalid command-line switch"
    sys.exit()

# stats on packets and keys received, and anomalies found

def stats():
  print >>fp
  print >>fp,"packets received = %d" % nrecv
  print >>fp,"unique keys = %d" % len(kv)
  print >>fp,"true anomalies = %d" % ptrue
  print >>fp,"false positives = %d" % pfalse
  print >>fp,"false negatives = %d" % nfalse
  print >>fp,"true negatives = %d" % ntrue

# process one packet

def packet(nvalues):
  global nrecv,nactive,nunique,ptrue,pfalse,nfalse,ntrue;

  nrecv += 1
  type,str,len = phish.unpack()
  
  # process all datums in packet

  lines = str.split('\n')
  
  for line in lines[1:-1]:
    words = line.split(',')

    # store datum in hash table
    # if new key, add to active set, deleting LRU key if necessary
    # count = # of times key has been seen
    # value = # of times value = 1
    # value < 0 means already seen Nthresh times

    key = int(words[0])
    value = int(words[1])

    if key not in kv:
      nunique += 1
      if nactive < maxactive:
        akey = aset[nactive]
        nactive += 1
        kv[key] = akey
      else:
        akey = list.last
        list.remove(akey)
        del kv[akey.key]
        kv[key] = akey
      akey.key = key
      akey.count = 1
      akey.value = value
      list.prepend(akey)

    else:
      akey = kv[key]
      list.move2front(akey)
      
      if akey.value < 0:
        akey.count += 1
        continue
      
      akey.count += 1
      akey.value += value
      
      if akey.count == nthresh:
        truth = int(words[2])
        if akey.value > mthresh:
          if truth:
            nfalse += 1
            print >>fp,"false negative =",key
          else: ntrue += 1
        else:
          if truth:
            ptrue += 1
            print >>fp,"true anomaly =",key
          else:
            pfalse += 1
            print >>fp,"false positive =",key
        akey.value = -1

# main program

args = phish.init(sys.argv)
phish.input(0,packet,None,1)
phish.check()

fp = open("tmp.%d" % phish.query("idlocal",0,0),'w');
options(args)

aset = maxactive*[0]

# key-value hash table for received datums
# aset = list of active keys

aset = maxactive*[0]
for i in xrange(maxactive): aset[i] = Akey()
list = MyDoubleLinkedList()
nactive = 0

phish.loop()

stats()
fp.close()
phish.exit()

File: analytics/anomaly2/phish/anomaly2.cpp
// MINNOW anomaly2
// anomaly detection from unbounded key range
// print found anomalies to screen

#ifndef __STDC_FORMAT_MACROS
#define __STDC_FORMAT_MACROS
#endif

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdexcept>

#include <tr1/unordered_map>

#include "phish.h"
#include "my_double_linked_list.h"

void packet(int);
void stats();
void options(int, char **);

/* ---------------------------------------------------------------------- */

// hard-coded settings from generator #2

#define NACTIVE 131072

// defaults for command-line switches

int nthresh = 24;
int mthresh = 4;
int nsenders = 1;
int maxactive = 2*NACTIVE;

// packet stats

uint64_t nrecv = 0;
uint64_t nunique = 0;

// anomaly stats

int ptrue = 0;
int pfalse = 0;
int nfalse = 0;
int ntrue = 0;

// one active key

struct Akey {
  uint64_t key;
  int count,value;
  Akey *prev;
  Akey *next;
};

// active set of keys

Akey *aset;
MyDoubleLinkedList<Akey*> list;
int nactive = 0;

// key-value hash table for received datums

std::tr1::unordered_map<uint64_t,Akey*> kv(1);

// output file handle

char *fname = NULL;
FILE *fp;

/* ---------------------------------------------------------------------- */

int main(int narg, char **args)
{
  phish_init(&narg,&args);
  phish_input(0,packet,NULL,1);
  phish_check();

  options(narg,args);

  // optional file output

  if (fname) {
    char fextend[16];
    sprintf(fextend,"%s.%d",fname,phish_query("idlocal",0,0));
    fp = fopen(fextend,"w");
  } else fp = stdout;

  // aset = active key set
  // stored as doubly-linked list so can delete least recently used

  aset = new Akey[nsenders*maxactive];
  //kv.reserve(maxactive);
  kv.rehash(ceil(nsenders*maxactive / kv.max_load_factor()));

  phish_loop();

  stats();
  fclose(fp);
  phish_exit();
}

/* ---------------------------------------------------------------------- */

void packet(int nvalues)
{
  char *buf,*datum;
  int len;
  Akey *akey;

  nrecv++;
  phish_unpack(&buf,&len);
  buf[len] = '\0';

  // scan past header line

  strtok(buf,"\n");

  // process variable # of datums in packet
  // ignore perpacket setting

  while (1) {
    if ((datum = strtok(NULL,",\n")) == NULL) break;
    uint64_t key = strtoul(datum,NULL,0);
    uint32_t value = strtoul(strtok(NULL,",\n"),NULL,0);
    uint32_t truth = strtoul(strtok(NULL,",\n"),NULL,0);

    // store datum in hash table
    // if new key, add to active set, deleting LRU key if necessary
    // count = # of times key has been seen
    // value = # of times value = 1
    // value < 0 means already seen Nthresh times
    
    if (!kv.count(key)) {
      nunique++;
      if (nactive < maxactive) {
        akey = &aset[nactive++];
        kv[key] = akey;
      } else {
        akey = list.last;
        list.remove(akey);
        kv.erase(akey->key);
        kv[key] = akey;
      }
      akey->key = key;
      akey->count = 1;
      akey->value = value;
      list.prepend(akey);

    } else {
      akey = kv[key];
      list.move2front(akey);

      if (akey->value < 0) {
        akey->count++;
        continue;
      }
      
      akey->count++;
      akey->value += value;

      if (akey->count == nthresh) {
        if (akey->value > mthresh) {
          if (truth) {
            nfalse++;
            fprintf(fp,"false negative = %" PRIu64 "\n",key);
          } else ntrue++;
        } else {
          if (truth) {
            ptrue++;
            fprintf(fp,"true anomaly = %" PRIu64 "\n",key);
          } else {
            pfalse++;
            fprintf(fp,"false positive = %" PRIu64 "\n",key);
          }
        }
        akey->value = -1;
      }
    }
  }
}

/* ---------------------------------------------------------------------- */

// stats on packets and keys received, and anomalies found

void stats()
{
  fprintf(fp,"\n");
  fprintf(fp,"packets received = %" PRIu64 "\n",nrecv);
  fprintf(fp,"unique keys = %" PRIu64 "\n",nunique);
  fprintf(fp,"true anomalies = %d\n",ptrue);
  fprintf(fp,"false positives = %d\n",pfalse);
  fprintf(fp,"false negatives = %d\n",nfalse);
  fprintf(fp,"true negatives = %d\n",ntrue);
}

/* ---------------------------------------------------------------------- */

// parse command line
// optional arguments:
//   -t = check key for anomalous when appears this many times
//   -m = key is anomalous if value=1 <= this many times
//   -p = # of generators sending to me, used to pre-allocate hash table
//   -a = size of active key set
//   -f = filename for output

void options(int narg, char **args)
{
  int op;
  
  while ( (op = getopt(narg,args,"t:m:p:a:f:")) != EOF) {
    switch (op) {
    case 't':
      nthresh = atoi(optarg);
      break;
    case 'm':
      mthresh = atoi(optarg);
      break;
    case 'p':
      nsenders = atoi(optarg);
      break;
    case 'a':
      maxactive = atoi(optarg);
      maxactive *= 2;
      break;
    case 'f':
      if (strcmp(optarg,"NULL") == 0) fname = NULL;
      else fname = optarg;
      break;
    }
  }

  if (optind != narg) phish_error("Syntax: anomaly2 options");
}
